smalldata_tools/ana_funcs/dropletCode/greedyguess.py

```diff
@@ -1,40 +1,56 @@
 import numpy as np
 #Greedy assignment of photons
 #from fitdrop import placephots
-#from numba import jit
+from numba import jit
 
-#@jit
+@jit(nopython=True, cache=True)
 def greedyguess(img,nophots,aduspphot):
-    #relative indices for the 4 tiles, order is carefully chosen
-    xs=np.array([[0,0],[0,0],[-1,0],[1,0]])
-    ys=np.array([[-1,0],[1,0],[0,0],[0,0]])
-
-    timg=img.copy()
     pxs=np.zeros(nophots)
     pys=pxs.copy()
-    for n in range(nophots):
+    
+    fimg=img.copy()/aduspphot
+    pimg=fimg//1
+    timg=fimg-pimg
+
+    nophots_assigned=0
+    #pimg->sparse array, fill pxs&pys, figure out what to do with photons > 1.
+    for ii in range(pimg.shape[0]):
+        for jj in range(pimg.shape[1]):
+            for photon in range(pimg[ii,jj]):
+                pxs[nophots_assigned] = ii
+                pys[nophots_assigned] = jj
+                nophots_assigned=nophots_assigned+1
+
+    for n in range(nophots_assigned,nophots):
         #find max intensity for next photon guess
-        i,j=np.unravel_index(np.argmax(timg),timg.shape)
+        maximg=-1
+        for ii in range(timg.shape[0]):
+            for jj in range(timg.shape[1]):
+                if timg[ii,jj]>maximg:
+                    i=ii
+                    j=jj
+                    maximg = timg[ii,jj]
+        #i,j=np.unravel_index(np.argmax(timg),timg.shape)
         #put a photon here if it fits
-        if(timg[i,j]>aduspphot):
-            timg[i,j] -= aduspphot
-            pxs[n] = i
-            pys[n] = j
-            continue #go to next photon
-        Is=np.sum(timg[i+xs,j+ys],axis=1)#sum the four pairs at (i,j)
-        #find max pair
-        mi=np.argmax(Is)
-        #put all of pixel i,j into photon (not C.O.M.)
-        t = 1.0-timg[i,j]/aduspphot  #fraction of photon not in i,j
+        t = 1.0-timg[i,j]  #fraction of photon not in i,j
         timg[i,j]=0.0 #use up central pixel
-        if(mi<2): #vertical pair
-            p1i=0.0
-            p1j=t*ys[mi,0] #vertical shift
-            timg[i,j+ys[mi,0]]=np.maximum(0.0,timg[i,j+ys[mi,0]]-t*aduspphot)
-        else: #horizontal pair
-            p1i=t*xs[mi,0] #horizontal shift
-            p1j=0.0
-            timg[i+xs[mi,0],j]=np.maximum(0.0,timg[i+xs[mi,0],j]-t*aduspphot)
-        pxs[n]=i+p1i
-        pys[n]=j+p1j
+
+        #find the max neighboring pixel
+        if timg[i-1,j]>max(timg[i+1,j],timg[i,j-1],timg[i,j+1]):
+            pxs[n]=i-t
```

Comment:
Am i understanding correctly that the photon position is assigned to a distance equal to the fractional photon intensity of the weaker pixel away from the more intense one? (not sure I make myself clear here). 
This trivially correspond to the center of mass between the two pixels then, is that right?

Reply:
Yes, again this is inherited from the original code (and maybe somewhat rewritten)